File: src/main/java/top/kmar/mi/api/net/handler/CommonMessage.java
package top.kmar.mi.api.net.handler;

import io.netty.buffer.ByteBuf;
import net.minecraft.entity.player.EntityPlayerMP;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraftforge.fml.common.network.ByteBufUtils;
import net.minecraftforge.fml.common.network.simpleimpl.IMessage;
import net.minecraftforge.fml.relauncher.Side;
import net.minecraftforge.fml.relauncher.SideOnly;
import top.kmar.mi.api.utils.expands.IOExpandsKt;

/**
 * 通用信息传输
 * @author EmptyDreams
 */
public class CommonMessage implements IMessage {
	
	private NBTTagCompound data;
	private String key;
	
	public CommonMessage(NBTTagCompound data, String key) {
		this.data = data;
		this.key = key;
	}
	
	public CommonMessage() { }
	
	/** 服务端解析数据 */
	public void parseServer(EntityPlayerMP player) {
		ServiceRawQueue.add(data, key, player);
	}
	
	/** 客户端解析数据 */
	@SideOnly(Side.CLIENT)
	public void parseClient() {
		ClientRawQueue.add(data, key);
	}
	
	@Override
	public void fromBytes(ByteBuf buf) {
		this.key = IOExpandsKt.readString(buf);
		data = ByteBufUtils.readTag(buf);
	}
	
	@Override
	public void toBytes(ByteBuf buf) {
		if (data == null) throw new IllegalArgumentException("没有给定需要传输的信息");
		IOExpandsKt.writeString(buf, key);
		ByteBufUtils.writeTag(buf, data);
	}
	
}

File: src/main/java/top/kmar/mi/api/net/CommonMessage.kt
package top.kmar.mi.api.net

import io.netty.buffer.ByteBuf
import net.minecraft.nbt.NBTBase
import net.minecraftforge.fml.common.network.simpleimpl.IMessage
import top.kmar.mi.api.utils.expands.readNbt
import top.kmar.mi.api.utils.expands.readString
import top.kmar.mi.api.utils.expands.writeNbt
import top.kmar.mi.api.utils.expands.writeString

/**
 * 通用信息
 * @author EmptyDreams
 */
class CommonMessage(
    key: String, data: NBTBase
) : IMessage {

    /** 处理该信息的处理器的 key */
    var key = key
        private set
    /** 要传递的信息 */
    var data = data
        private set

    override fun fromBytes(buf: ByteBuf) {
        key = buf.readString()
        data = buf.readNbt()
    }

    override fun toBytes(buf: ByteBuf) {
        buf.writeString(key)
        buf.writeNbt(data)
    }

}